contracts-periphery/src/WithdrawHook.sol

```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+pragma abicoder v2;
+
+import "openzeppelin-contracts/contracts/token/ERC20/IERC20.sol";
+import "openzeppelin-contracts/contracts/token/ERC20/utils/SafeERC20.sol";
+import "src/utils/ISwapRouter.sol";
+
+contract WithdrawHook {
+
+  ISwapRouter internal immutable swapRouter;
+  address internal constant WETH9 = 0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2;
+
+  constructor(ISwapRouter _swapRouter) {
+  swapRouter = _swapRouter;
+  }
+
+  function tokensWithdrawn(
+    uint256 _amount,
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    bytes memory _data
+  ) external{
+    // Plz recommend checks that should happen here
+    _swapForEth(_amount, _stealthAddr, _acceptor, _tokenAddr, _sponsor, _sponsorFee, _data);
+  }
+
+  function _swapForEth(
+    uint256 _amount,
+    address _stealthAddr,
+    address _acceptor,
+    address _tokenAddr,
+    address _sponsor,
+    uint256 _sponsorFee,
+    bytes memory _data) internal {
+    uint256 senderBalance = _amount;
+    if(_sponsorFee > 0) {
+      SafeERC20.safeTransferFrom(IERC20(_tokenAddr), address(this), _sponsor, _sponsorFee);
+      senderBalance -= _sponsorFee;
+    }
+
+    (address _dest, uint256 _swapAmount, uint256 _amountOutMinimum, uint24 _poolFee) = abi.decode(_data, (address, uint256, uint256, uint24));
+    senderBalance -= _swapAmount;
+
+    SafeERC20.safeApprove(IERC20(_tokenAddr), address(swapRouter), _swapAmount);
+
+    ISwapRouter.ExactInputSingleParams memory params =
+      ISwapRouter.ExactInputSingleParams({
+        tokenIn: _tokenAddr,
+        tokenOut: WETH9,
```

Comment:
Why do we hardcode ETH as the output here? As mentioned in the spec comment, I think both of these should be input parameters. This approach prevents someone from e.g. swapping ETH to UDSC and sending it directly to an exchange address.

Reply:
The withdraw hook would never be used by a user withdrawing ETH, since that ETH is already in an EOA, hooks don't apply.

The use case imagined here is specifically being able to swap some of your tokens into ETH and send it all to a freshly generated address, thus avoiding the need to fund that address before using it. You could argue we could make it more flexible by allowing a swap to any token, but I don't really see a use case for that, and it does add a small amount of gas and complexity.

I've documented the imagined usecase in an issue here: https://github.com/ScopeLift/umbra-protocol/issues/345

Reply:
That's true, if a user received ETH they can just use the regular SwapRouter. But hardcoding ETH as an output here still prevents you from e.g. receiving DAI and swapping half for ETH and half for USDT, and sending ETH and USDT to the new address. That scenario is not mentioned in #345

Reply:
I would argue we really shouldn't be worrying about that scenario, because as long as the user can swap some of the token for ETH, they can then use the address afterwards to do whatever they want (including swapping DAI to USDT). I don't think we should take on complexity in our contract just to spare the user a transaction in a hypothetical use case.

Reply:
My proposal for how to implement this is here: https://github.com/ScopeLift/umbra-protocol/pull/341#discussion_r890318294

IMO that approach is better than hardcoding ETH because it's more flexible, still limits you to a single swap, but also supports withdrawing arbitrary tokens to an exchange such as "swap DAI for USDC and send it to an exchange"

Reply:
Token outputs, swap paths, are now passed in as multicall data https://github.com/ScopeLift/umbra-protocol/pull/341/commits/b40d4a8da174dc150d24e62c76a20189c2d8cfc2